File: app/macos/hyperwhisper/Services/ConfigService.swift
//
//  ConfigService.swift
//  hyperwhisper
//
//  Fetches remote trial configuration from the server.
//  Caches values in UserDefaults with server-driven expiry (Cache-Control max-age).
//  Falls back to hardcoded defaults when offline or on error.
//

import Foundation
import os

struct RemoteConfig {
    let trialDailyLimitSeconds: Int
    let trialModelDownloadLimit: Int
    /// Server-driven cache TTL (Cache-Control max-age, seconds). Threaded into the
    /// core's remote-override store so `licenseRemoteOverrideIfFresh` honors it (B4).
    let maxAgeSeconds: Int
}

final class ConfigService {

    static let shared = ConfigService()

    private let logger = Logger(subsystem: "com.hyperwhisper.hyperwhisper", category: "ConfigService")

    // UserDefaults keys
    private let kDailyLimit = "config.trialDailyLimitSeconds"
    private let kModelLimit = "config.trialModelDownloadLimit"
    private let kLastFetch = "config.lastFetchTimestamp"
    private let kMaxAge = "config.maxAge"

    private init() {}

    /// Returns cached config if fresh (within server's max-age), otherwise nil.
    func getCachedConfig() -> RemoteConfig? {
        let defaults = UserDefaults.standard

        guard defaults.object(forKey: kDailyLimit) != nil,
              defaults.object(forKey: kLastFetch) != nil else {
            return nil
        }

        let lastFetch = defaults.double(forKey: kLastFetch)
        let maxAge = defaults.double(forKey: kMaxAge)

        // Check if cache is still fresh
        if Date().timeIntervalSince1970 - lastFetch > maxAge {
            return nil
        }

        return RemoteConfig(
            trialDailyLimitSeconds: defaults.integer(forKey: kDailyLimit),
            trialModelDownloadLimit: defaults.integer(forKey: kModelLimit),
            maxAgeSeconds: Int(maxAge)
        )
    }

    /// Fetches config from the server, caches it, and returns the values.
    /// Returns nil on any error (network, parse, non-200 status).
    func fetchConfig() async -> RemoteConfig? {
        let urlString = "\(NetworkConfig.baseURL)/api/config"
        guard let url = URL(string: urlString) else { return nil }

        do {
            var request = URLRequest(url: url)
            request.httpMethod = "GET"
            request.timeoutInterval = 10
            request.setValue("HyperWhisper/1.0", forHTTPHeaderField: "User-Agent")

            let (data, response) = try await URLSession.shared.data(for: request)

            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                logger.warning("Config fetch failed with non-200 status")
                return nil
            }

            guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let dailyLimit = json["trial_daily_limit_seconds"] as? Int,
                  let modelLimit = json["trial_model_download_limit"] as? Int else {
                logger.warning("Config fetch: failed to parse response")
                return nil
            }

            // Parse max-age from Cache-Control header
            var maxAge: Double = 21600 // default 6 hours
            if let cacheControl = httpResponse.value(forHTTPHeaderField: "Cache-Control") {
                if let range = cacheControl.range(of: "max-age=") {
                    let afterMaxAge = cacheControl[range.upperBound...]
                    if let seconds = Double(afterMaxAge.prefix(while: { $0.isNumber })) {
                        maxAge = seconds
                    }
                }
            }

            // Cache to UserDefaults
            let defaults = UserDefaults.standard
            defaults.set(dailyLimit, forKey: kDailyLimit)
            defaults.set(modelLimit, forKey: kModelLimit)
            defaults.set(Date().timeIntervalSince1970, forKey: kLastFetch)
            defaults.set(maxAge, forKey: kMaxAge)

            logger.info("Config fetched: dailyLimit=\(dailyLimit, privacy: .public)s, modelLimit=\(modelLimit, privacy: .public), maxAge=\(maxAge, privacy: .public)s")

            return RemoteConfig(
                trialDailyLimitSeconds: dailyLimit,
                trialModelDownloadLimit: modelLimit,
                maxAgeSeconds: Int(maxAge)
            )
        } catch {
            logger.warning("Config fetch error: \(error.localizedDescription, privacy: .public)")
            return nil
        }
    }
}

File: app/windows/HyperWhisper/Services/ConfigService.cs
// CONFIG SERVICE
// Fetches remote trial configuration from the server.
//
// TODO-verify (Windows/CI): Rust shared-core swap. UNVERIFIED / compile-only.
//
// Wave 3 swap: the remote trial-limit OVERRIDE (value + freshness TTL) is now
// owned by the `hw-license` Rust core, persisted via RustCoreKeyValueStore under
// com.hyperwhisper.config.*. This service keeps ONLY the HTTP GET of /api/config
// (the core has no config endpoint); the parsed values are handed to the core
// via LicenseStoreRemoteOverride, and GetCachedConfig reads them back via
// LicenseRemoteOverrideIfFresh.
//
// TTL: server-driven (B4). FetchConfigAsync writes the response's Cache-Control
// max-age into the store (key "com.hyperwhisper.config.maxAgeSecs"); the core's
// freshness check honors it, defaulting to 6h when absent and clamping to a 24h
// upper bound (REMOTE_OVERRIDE_TTL_SECS). Restores the pre-unification behavior.

using System;
using System.Net.Http;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;
using uniffi.hyperwhisper_core;
using HttpMethod = System.Net.Http.HttpMethod;

namespace HyperWhisper.Services;

/// <summary>
/// Fetches remote trial configuration; caches it through the Rust core.
/// </summary>
public sealed class ConfigService
{
    // =========================================================================
    // SINGLETON
    // =========================================================================

    private static ConfigService? _instance;
    private static readonly object _lock = new();

    public static ConfigService Instance
    {
        get
        {
            if (_instance == null)
            {
                lock (_lock)
                {
                    _instance ??= new ConfigService();
                }
            }
            return _instance;
        }
    }

    // =========================================================================
    // CONSTANTS
    // =========================================================================

    private const string ConfigUrl = "https://www.hyperwhisper.com/api/config";

    // =========================================================================
    // STATE
    // =========================================================================

    private static readonly HttpClient _httpClient = new()
    {
        Timeout = TimeSpan.FromSeconds(10)
    };

    private ConfigService() { }

    // =========================================================================
    // PUBLIC API
    // =========================================================================

    /// <summary>
    /// Returns the cached remote override if still fresh (24h TTL, in the core),
    /// otherwise null.
    /// </summary>
    public RemoteConfig? GetCachedConfig()
    {
        // TODO-verify (Windows/CI): Rust shared-core swap.
        TrialLimits? over = HyperwhisperCoreMethods.LicenseRemoteOverrideIfFresh(
            RustLicenseCore.Store, RustLicenseCore.Now());
        if (over == null)
        {
            return null;
        }
        return new RemoteConfig((int)over.dailySeconds, (int)over.modelDownloads);
    }

    /// <summary>
    /// Fetches config from the server, persists it through the core, and returns
    /// the values. Returns null on any error.
    /// </summary>
    public async Task<RemoteConfig?> FetchConfigAsync()
    {
        try
        {
            var request = new HttpRequestMessage(HttpMethod.Get, ConfigUrl);
            request.Headers.Add("User-Agent", "HyperWhisper/1.0");

            var response = await _httpClient.SendAsync(request);
            if (!response.IsSuccessStatusCode)
            {
                LoggingService.Warn($"ConfigService: Fetch failed with status {response.StatusCode}");
                return null;
            }

            var body = await response.Content.ReadAsStringAsync();
            var parsed = JsonSerializer.Deserialize<ConfigResponse>(body);
            if (parsed == null)
            {
                LoggingService.Warn("ConfigService: Failed to parse response");
                return null;
            }

            // Persist the override through the core.
            // TODO-verify (Windows/CI): Rust shared-core swap.
            HyperwhisperCoreMethods.LicenseStoreRemoteOverride(
                RustLicenseCore.Store,
                new TrialLimits(
                    @dailySeconds: parsed.TrialDailyLimitSeconds,
                    @modelDownloads: parsed.TrialModelDownloadLimit),
                RustLicenseCore.Now());

            // Persist the server's Cache-Control max-age so the core's freshness
            // check honors it (B4). Key string must match hw_license::cache's
            // K_OVERRIDE_MAX_AGE ("com.hyperwhisper.config.maxAgeSecs"). Absent /
            // non-positive ⇒ the core falls back to its 6h default, clamped to 24h.
            var maxAge = response.Headers.CacheControl?.MaxAge;
            var maxAgeSecs = maxAge.HasValue ? (long)maxAge.Value.TotalSeconds : 0;
            if (maxAgeSecs > 0)
            {
                RustLicenseCore.Store.Set(
                    "com.hyperwhisper.config.maxAgeSecs",
                    maxAgeSecs.ToString());
            }
            else
            {
                // FetchConfigAsync is always a LIVE fetch. With no/≤0 max-age on
                // this response, CLEAR any previously stored value so the core's
                // effective_override_ttl reverts to its 6h default (it treats an
                // empty/unparseable value as "use default") instead of pinning a
                // stale TTL from an earlier response that carried max-age.
                RustLicenseCore.Store.Set("com.hyperwhisper.config.maxAgeSecs", "");
            }

            LoggingService.Info(
                $"ConfigService: Fetched config (dailyLimit={parsed.TrialDailyLimitSeconds}s, modelLimit={parsed.TrialModelDownloadLimit})");

            return new RemoteConfig(parsed.TrialDailyLimitSeconds, parsed.TrialModelDownloadLimit);
        }
        catch (Exception ex)
        {
            LoggingService.Warn($"ConfigService: Fetch error: {ex.Message}");
            return null;
        }
    }

    // =========================================================================
    // DATA MODELS
    // =========================================================================

    private class ConfigResponse
    {
        [JsonPropertyName("trial_daily_limit_seconds")]
        public int TrialDailyLimitSeconds { get; set; }

        [JsonPropertyName("trial_model_download_limit")]
        public int TrialModelDownloadLimit { get; set; }
    }
}

/// <summary>
/// Immutable remote config values.
/// </summary>
public sealed class RemoteConfig
{
    public int TrialDailyLimitSeconds { get; }
    public int TrialModelDownloadLimit { get; }

    public RemoteConfig(int trialDailyLimitSeconds, int trialModelDownloadLimit)
    {
        TrialDailyLimitSeconds = trialDailyLimitSeconds;
        TrialModelDownloadLimit = trialModelDownloadLimit;
    }
}
